src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1050,23 +1051,92 @@ StateForCase makeNext(Tests remainingTests)
             resultForRelation(BinaryOperatorKind relation, ConstantValue value)
             {
                 var input = test.Input;
-                IValueSetFactory? valueFac = ValueSetFactory.ForType(input.Type);
+                IValueSetFactory? valueFac = ValueSetFactory.ForInput(input);
                 if (valueFac == null || value.IsBad)
                 {
                     // If it is a type we don't track yet, assume all values are possible
                     return (values, values, true, true);
                 }
                 IValueSet fromTestPassing = valueFac.Related(relation.Operator(), value);
                 IValueSet fromTestFailing = fromTestPassing.Complement();
-                if (values.TryGetValue(test.Input, out IValueSet? tempValuesBeforeTest))
+                if (values.TryGetValue(input, out IValueSet? tempValuesBeforeTest))
                 {
                     fromTestPassing = fromTestPassing.Intersect(tempValuesBeforeTest);
                     fromTestFailing = fromTestFailing.Intersect(tempValuesBeforeTest);
                 }
+                bool truePossible = !fromTestPassing.IsEmpty;
+                bool falsePossible = !fromTestFailing.IsEmpty;
+                if (relation == BinaryOperatorKind.Equal &&
+                    input.Source is BoundDagPropertyEvaluation { IsLengthOrCount: true } s)
+                {
+                    Debug.Assert(value.Discriminator == ConstantValueTypeDiscriminator.Int32);
+                    int lengthValue = value.Int32Value;
+                    int commonIndex = -1; // common index from start, -1 means we don't have any
+                    IValueSet? commonValues = null;
+                    var temps = ArrayBuilder<BoundDagTemp>.GetInstance();
+                    foreach (var (temp, remainingValues) in values)
+                    {
+                        if (temp.Source is BoundDagIndexerEvaluation e)
+                        {
+                            (BoundDagTemp canonicalInput, BoundDagTemp lengthTemp, int index) = GetCanonicalInput(e);
+                            if (input.Equals(lengthTemp))
+                            {
+                                Debug.Assert(s.Input.Equals(canonicalInput));
+                                int normalizedIndex = index < 0 ? index + lengthValue : index;
+                                if (commonIndex < 0)
+                                {
+                                    (commonValues, commonIndex) = (remainingValues, normalizedIndex);
+                                    temps.Add(temp);
+                                }
+                                else if (commonIndex == normalizedIndex)
+                                {
+                                    commonValues = commonValues!.Intersect(remainingValues);
+                                    temps.Add(temp);
```

Comment:
If understood this section correctly, it looks like we find one pair of temps with same normalized index, then intersect their values and use that to update both in the main dictionary. 
But I would have expected we'd need to update more than one pair in some situations. 

<details><summary>Example</summary>
```
        [Fact]
        public void Subsumption_TODO2()
        {
            var src = @"
class C
{
    void Test(int[] a)
    {
        _ = a switch
        {
            [.., >0, > 0] => 0,
            [<= 0, <=0, ..] => 1,
            [var unreachable, var unreachable2] => 3,
            _  => 4,
        };
    }
}" + TestSources.GetSubArray;
            var comp = CreateCompilationWithIndexAndRange(src);
            comp.VerifyEmitDiagnostics(); // missing diagnostic
        }
```
</details>

Reply:
Although that example showed a problem, it's not the problem I had in mind (doesn't reach this new block of code).

Here's a variation that involves tracking remaining values:
<details><summary>Example</summary>
```
        [Fact]
        public void Subsumption_TODO2()
        {
            var src = @"
class C
{
    void Test(int[] a)
    {
        _ = a switch
        {
            [.., >0, >0] => 0,
            [0, 0, ..] => 1,
            [<0, <0, ..] => 1,
            [var unreachable, var unreachable2] => 3,
            _  => 4,
        };
    }
}" + TestSources.GetSubArray;
            var comp = CreateCompilationWithIndexAndRange(src);
            comp.VerifyEmitDiagnostics(
                // (6,15): warning CS8509: The switch expression does not handle all possible values of its input type (it is not exhaustive). For example, the pattern '_' is not covered.
                //         _ = a switch
                Diagnostic(ErrorCode.WRN_SwitchExpressionNotExhaustive, "switch").WithArguments("_").WithLocation(6, 15)
                ); // unexpected diagnostic
        }
````
</details>

Reply:
Corrected last example to account for four quadrants. I hope I got it right this time.